[Docs] Translate data_transform.md to English (#2325)
open-mmlab/mmcv: docs/en/understand_mmcv/data_transform.md

# Data Transformation

> In the OpenMMLab algorithm library, dataset construction and data preparation are decoupled. Usually, the construction of the data set only parses the dataset and records the basic information of each sample. The data preparation is a series of data transformations, data loading, preprocessing, formatting, and other operations are performed according to the basic information of the sample.

Reviewer: ```suggestion
In the OpenMMLab algorithm library, dataset construction and data preparation are decoupled. Usually, the construction of the dataset only parses the dataset and records the basic information of each sample, while the data preparation is a series of data transformations including data loading, preprocessing, formatting, and other operations performed according to the basic information of the sample.
```

## Design of data transformation

> In MMCV, we use various callable data transformation classes to manipulate data. These data transformation classes can accept several configuration parameters for the instantiation and then process the input data dictionary by calling. All data transformation methods accept a dictionary as the input and output the processed data as a dictionary as well. A simple example is as follows:

Reviewer: ```suggestion
In MMCV, we use various callable data transformation classes to manipulate data. These data transformation classes can accept several configuration parameters for the instantiation and then process the input data dictionary by `__call__` method. All data transformation methods accept a dictionary as the input and produce the output as a dictionary as well. A simple example is as follows:
```

>>>
>>> transform = Resize(scale=(224, 224))
>>> data_dict = {'img': np.random.rand(256, 256, 3)}
>>> data_dict = transform(data_dict)
>>> print(data_dict['img'].shape)
(224, 224, 3)
```

The data transformation class reads some fields of the input dictionary and may add or update some fields. The keys of these fields are mostly fixed. For example, `Resize` will always read fields such as `"img"` in the input dictionary. More information about the conventions for input and output fields could be found in the documentation of the corresponding class.

MMCV provides a unified base class called `BaseTransform` for all data transformation classes:

```python
class BaseTransform(metaclass=ABCMeta):

    def __call__(self, results: dict) -> dict:

        return self.transform(results)

    @abstractmethod
    def transform(self, results: dict) -> dict:
        pass
```

All data transformation classes must inherit `BaseTransform` and implement the `transform` method. Both the input and output of the `transform` method are a dictionary. In the **Custom data transformation class** section, we will describe how to implement a data transformation class in more detail.

## Data pipeline

> As mentioned above, the input and output of all data transformations are a dictionary, and according to the \[Convention on Datasets\] (TODO) in OpenMMLab, the basic information of each sample in the dataset is a dictionary. This way, we can connect all data transformation operations end to end and combine them into a data pipeline. This pipeline inputs the information dictionary of the samples in the data set and outputs the information dictionary after a series of processing.

Reviewer: ```suggestion
As mentioned above, the inputs and outputs of all data transformations are dictionaries. Moreover, according to the \[Convention on Datasets\] (TODO) in OpenMMLab, the basic information of each sample in the dataset is also a dictionary. This way, we can connect all data transformation operations end to end and combine them into a data pipeline. This pipeline inputs the information dictionary of the samples in the dataset and outputs the information dictionary after a series of processing.
```

> Taking the classification task as an example, we show a typical data pipeline in the figure below. For each sample, the information stored in the data set is a dictionary, as shown on the far left in the figure. After each data transformation operation represented by the blue block, a new field (marked in green) will be added to the data dictionary or update an existing field (marked in orange).

Reviewer: ```suggestion
Taking the classification task as an example, we show a typical data pipeline in the figure below. For each sample, the information stored in the dataset is a dictionary, as shown on the far left in the figure. After each data transformation operation represented by the blue block, a new field (marked in green) will be added to the data dictionary or an existing field (marked in orange) will be updated.
```

<div align=center>
<img src="https://user-images.githubusercontent.com/26739999/154197953-bf0b1a16-3f41-4bc7-9e67-b2b9b323d895.png" width="90%"/>
</div>

The data pipeline is a list of several data transformation configuration dictionaries in the configuration file. Each dataset needs to set the parameter `pipeline` to define the data preparation operations the dataset needs to perform. The configuration of the above data pipeline in the configuration file is as follows:

```python
pipeline = [
    dict(type='LoadImageFromFile'),
    dict(type='Resize', size=256, keep_ratio=True),
    dict(type='CenterCrop', crop_size=224),
    dict(type='Normalize', mean=[123.675, 116.28, 103.53], std=[58.395, 57.12, 57.375]),
    dict(type='ClsFormatBundle')
]

dataset = dict(
    ...
    pipeline=pipeline,
    ...
)
```

## Common data transformation classes

> The commonly used data transformation classes can be roughly divided into data loading, data preprocessing and enhancement, and data formatting. In MMCV, we provide some commonly used classes as follows:

Reviewer: ```suggestion
The commonly used data transformation classes can be roughly divided into data loading, data preprocessing and augmentation, and data formatting. In MMCV, we provide some commonly used classes as follows:
```

### Data loading

To support the loading of large-scale datasets, data is usually not loaded when `Dataset` is initialized. Only the corresponding path is loaded. Therefore, it is necessary to load specific data in the data pipeline.

|            Class            |                    Feature                     |
| :-------------------------: | :--------------------------------------------: |
| [`LoadImageFromFile`](TODO) |              Load from file path               |
|  [`LoadAnnotations`](TODO)  | Load and organize the annotations (bbox, etc.) |

### Data preprocessing and enhancement

Data preprocessing and augmentation usually involve transforming the image itself, such as cropping, padding, scaling, etc.

|              Class               |                        Feature                         |
| :------------------------------: | :----------------------------------------------------: |

> |          [`Pad`](TODO)           |                        Padding                         |

Reviewer 2: Can we fill these empty links now?

Author: > Can we fill these empty links now?

I checked the mmcv's API pages, and it seems multiple features have already been finished. However, would the internal team take a look at these? I believe this would be more appropriate. @zhouzaida

Reviewer 3: Hi @xin-li-67 , it would be better to update the empty links. They can be found at https://mmcv.readthedocs.io/en/2.x/api/transforms.html.

Reviewer 2: > Hi @xin-li-67 , it would be better to update the empty links. They can be found at https://mmcv.readthedocs.io/en/2.x/api/transforms.html.

I think it's better to use doc anchors instead of links. If @xin-li-67 is not familiar with doc anchors, we can add it directly.

Author: > > Hi @xin-li-67 , it would be better to update the empty links. They can be found at https://mmcv.readthedocs.io/en/2.x/api/transforms.html.
> 
> I think it's better to use doc anchors instead of links. If @xin-li-67 is not familiar with doc anchors, we can add it directly.

Thanks for the comment. I'm not quite familiar with doc anchors indeed. Would you and the team please finish them directly? That will be grateful.

|       [`CenterCrop`](TODO)       |                      Center crop                       |
|       [`Normalize`](TODO)        |                  Image normalization                   |
|         [`Resize`](TODO)         |         Resize to the specified size or ratio          |
|      [`RandomResize`](TODO)      |  Scale the image randomly within the specified range   |
| [`RandomMultiscaleResize`](TODO) | Scale the image to a random size from multiple options |

> |    [`RandomGrayscale`](TODO)     |                    Random greyscale                    |

Reviewer 3: ```suggestion
|    [`RandomGrayscale`](TODO)     |                    Random grayscale                    |
```

|       [`RandomFlip`](TODO)       |                      Random flip                       |
|   [`MultiScaleFlipAug`](TODO)    |    Support scaling and flipping during the testing     |

### Data formatting

> Data formatting operations are typed conversions performed on the data.

Reviewer: ```suggestion
Data formatting operations are type conversions performed on the data.
```

|          Class          |                   Feature                    |
| :---------------------: | :------------------------------------------: |
|   [`ToTensor`](TODO)    | Convert the specified data to `torch.Tensor` |
| [`ImageToTensor`](TODO) |     Convert the image to `torch.Tensor`      |

> ## Customizing data transformation classes

Reviewer 3: ```suggestion
## Customize data transformation classes
```

> To implement a new data transformation class, you must extend `BaseTransform` and implement the `transform` method. Here, we use a simple flip transform (`MyFlip`) as an example:

Reviewer 3: ```suggestion
To implement a new data transformation class, you must inherit `BaseTransform` and implement the `transform` method. Here, we use a simple flip transform (`MyFlip`) as an example:
```

```python
import random
import mmcv
from mmcv.transforms import BaseTransform, TRANSFORMS

@TRANSFORMS.register_module()
class MyFlip(BaseTransform):
    def __init__(self, direction: str):
        super().__init__()
        self.direction = direction

    def transform(self, results: dict) -> dict:
        img = results['img']
        results['img'] = mmcv.imflip(img, direction=self.direction)
        return results
```

Now, we can instantiate `MyFlip` as a callable object to handle our data dictionary.

```python
import numpy as np

transform = MyFlip(direction='horizontal')
data_dict = {'img': np.random.rand(224, 224, 3)}
data_dict = transform(data_dict)
processed_img = data_dict['img']
```

Alternatively, use `MyFlip` transform in the `pipeline` of the config file.

```python
pipeline = [
    ...
    dict(type='MyFlip', direction='horizontal'),
    ...
]
```

It should be noted that if you want to use it in the configuration file, you must ensure that the file where the `MyFlip` class is located can be imported at the runtime.

## Transform wrapper

> Transform wrappers are a special class of data transformations. They do not operate on images, labels, and other information in the data dictionary by themselves, however, they enhance the behavior of data transformations defined in them.

Reviewer: ```suggestion
Transform wrappers are a special class of data transformations. They do not operate on images, labels or other information in the data dictionary by themselves. Instead, they enhance the behavior of data transformations defined in them.
```

### KeyMapper

> `KeyMapper` is used to map fields in the data dictionary. For example, usually image processing transforms get their values from the `"img"` field in the data dictionary. But sometimes we want these transforms to handle images in other fields in the data dictionary, such as the `"gt_img"` field.

Reviewer: ```suggestion
`KeyMapper` is used to map fields in the data dictionary. For example, image processing transforms usually get their values from the `"img"` field in the data dictionary. But sometimes we want these transforms to handle images in other fields in the data dictionary, such as the `"gt_img"` field.
```

When used with registry and configuration file, the field map wrapper should be used as follows:

```python
pipeline = [
    ...
    dict(type='KeyMapper',
        mapping={
            'img': 'gt_img',  # map "gt_img" to "img"
            'mask': ...,  # The "mask" field in the raw data is not used. That is, for wrapped data transformations, the "mask" field is not included in the data
        },
        auto_remap=True,  # remap "img" back to "gt_img" after the transformation
        transforms=[
            # only need to specify "img" in `RandomFlip`
            dict(type='RandomFlip'),
        ])
    ...
]
```

With `KeyMapper`, we don't need to consider various possible input field names in the `transform` method when we implement the data transformation class. We only need to deal with the default fields.

### RandomChoice and RandomApply

> `RandomChoice` is used to randomly apply a combination of data transformations from a series of combinations of data transformations. With this wrapper, we can easily implement some data augmentation functions, such as AutoAugment.

Reviewer: ```suggestion
`RandomChoice` is used to randomly select a data transformation pipeline from the given choices. With this wrapper, we can easily implement some data augmentation functions, such as AutoAugment.
```

> If used with the registrar and configuration file, you can use `RandomChoice` as the following example in the `pipeline` of the dataset in the configuration file:

Reviewer: ```suggestion
In configuration file, you can use `RandomChoice` as follows:
```

```python
pipeline = [
    ...
    dict(type='RandomChoice',
        transforms=[
            [
                dict(type='Posterize', bits=4),
                dict(type='Rotate', angle=30.)
            ],  # the first combo option
            [
                dict(type='Equalize'),
                dict(type='Rotate', angle=30)
            ],  # the second combo option
        ],
        prob=[0.4, 0.6]  # the prob of each combo
        )
    ...
]
```

`RandomApply` is used to randomly perform a combination of data transformations with a specified probability. For example:

```python
pipeline = [
    ...
    dict(type='RandomApply',
        transforms=[dict(type='Rotate', angle=30.)],
        prob=0.3)  # perform the transformation with prob as 0.3
    ...
]
```

### TransformBroadcaster

> Usually, a data transformation class only reads the target of an operation from one field. While we can also use `KeyMapper` to change the fields read, there is no way to apply transformations to the data of multiple fields at once. To achieve this, we need to use the multi-target extension wrapper (`TransformBroadcaster`).

Reviewer 3: ```suggestion
Usually, a data transformation class only reads the target of an operation from one field. While we can also use `KeyMapper` to change the fields read, there is no way to apply transformations to the data of multiple fields at once. To achieve this, we need to use the multi-target extension wrapper `TransformBroadcaster`.
```

> `TransformBroadcaster` It has two uses, one is to apply data transformation to multiple specified fields, and the other is to apply data transformation to a group of targets under a field.

Reviewer: ```suggestion
`TransformBroadcaster` has two uses, one is to apply data transformation to multiple specified fields, and the other is to apply data transformation to a group of targets under a field.
```

           # more introduction will be referred in the following chapters (random variable sharing)
           share_random_params=True,
           transforms=[
               # "img" and "img_shape" fields are manipulated in the `RandomCrop` class
               # if "img_shape" is missing, only operate on "img"
               dict(type='RandomCrop'),
           ])
   ]
   ```

2. A set of targets applied to a field

   Suppose we need to apply a data transformation to the `"images"` field, which is a list of images.

   ```python
   pipeline = [
       dict(type='TransformBroadcaster',
           # map each image under the "images" field to the "img" field
           mapping={'img': 'images'},
           # remap the images under the "img" field back to the list in the "images" field after the transformation
           auto_remap=True,
           # whether to share random variables in the transformation of each target
           share_random_params=True,
           transforms=[
               # in the `RandomFlip` transformation class, we only need to manipulate the "img" field
               dict(type='RandomFlip'),
           ])
   ]
   ```

#### Decorator `cache_randomness`

> In `TransformBroadcaster`, we provide the `share_random_params` option to support sharing random states across multiple data transformations. For example, in a super-resolution task, we want to apply random transformations **synchronously** to the low-resolution image and the original image. If we use this function in a custom data transformation class, we need to mark which random variables support sharing in the class. This can be achieved with the decorator `cache_randomness`.

Reviewer: ```suggestion
In `TransformBroadcaster`, we provide the `share_random_params` option to support sharing random states across multiple data transformations. For example, in a super-resolution task, we want to apply **the same** random transformations **simultaneously** to the low-resolution image and the original image. If we use this function in a custom data transformation class, we need to mark which random variables support sharing in the class. This can be achieved with the decorator `cache_randomness`.
```
